src/components/Profile/useUserInfoForm.ts

```diff
@@ -0,0 +1,109 @@
+import { useCallback, useMemo, useState } from 'react';
+import { USERNAME_REGEX } from 'utils/regex';
+import { pause } from 'utils/time';
+
+import { BIO_MAX_CHAR_COUNT } from './UserInfoForm';
+
+type Props = {
+  onSuccess: () => void;
+};
+
+export default function useUserInfoForm({ onSuccess }: Props) {
+  // TODO__v1: auto-populate username and bio from useSwr()
+  // const { username, bio } = useSwr('/...')
+
+  const [username, setUsername] = useState('');
+  const [usernameError, setUsernameError] = useState('');
+
+  const [bio, setBio] = useState('');
+
+  // generic error that doesn't belong to username / bio
+  const [generalError, setGeneralError] = useState('');
+
+  const handleCreateUser = useCallback(async () => {
+    setGeneralError('');
+
+    //------------ client-side checks ------------
+    if (!username.length) {
+      setUsernameError('Required.');
+      return;
+    }
+
+    if (username.length < 2 || username.length > 20) {
+      setUsernameError('Username must be between 2 and 20 characters.');
```

Comment:
Maybe we can put these error messages in a single const like a map

Reply:
Agreed. In an ideal world we have a set of Validators that run against the input, and each validator returns a specific error const.

Example array of validators: [required, minLength, maxLength, ...] then run your input through them all and return first error (or if no error return true)